### ISSUE TITLE
Soap WS integration and soap connector problem

### ISSUE THREAD

**Issue by marioestradarosa** [strongloop/loopback-next#1553]
## Description

I created a sample project that interacts with a calculator SOAP web service, which exposes 4 methods. I bind the provider class using this.bind `this.bind(key).toProvider(provider);`  in the Application.ts file. The service is implementing Provider properly and the endpoints can be access with POST methods and in the body section I pass over the two parameter in a JSON object. It is also injected in the controller.

## Current Behavior

Whenever I run `npm start` and try to access any of the exposed endpoints the very first time, I get error 500 Type Error: The this.calculatorService.method is not a function. However, if I hit the send button again (I am using postman) it works fine, even with the other end points.

## Expected Behavior

The calculatorService, should be instantiated as the code in the BootMixin implementation dictates and users should see the result without any error since the first time. I am not sure if it has to do with LoopBack itself or [My code](https://github.com/marioestradarosa/loopback-soap-example) , please review and provide feedback!

<img width="1058" alt="screen shot 2018-07-18 at 23 52 55" src="https://user-images.githubusercontent.com/4633823/42925360-2cc73d32-8aeb-11e8-8743-358faab27eb3.png">

<img width="945" alt="screen shot 2018-07-18 at 23 53 06" src="https://user-images.githubusercontent.com/4633823/42925391-44831324-8aeb-11e8-8d0e-50009559c533.png">

And finally by pressing the Send button again, it works

<img width="1054" alt="screen shot 2018-07-18 at 23 53 31" src="https://user-images.githubusercontent.com/4633823/42925406-4fc0a472-8aeb-11e8-84c0-dc0bc96b8fe3.png">

_See [Reporting Issues](http://loopback.io/doc/en/contrib/Reporting-issues.html) for more tips on writing good issues_

**Comment by marioestradarosa** [comment]
**Follow up**: This only happens with the soap connector.
Here is a the output from debug:

Whenever the endpoint is reached (ie:  _http://localhost:3000/add/50/20_ ) 

<img width="1067" alt="screen shot 2018-07-23 at 22 25 15" src="https://user-images.githubusercontent.com/4633823/43116912-51d83402-8ec7-11e8-91ba-57490008fc0c.png">

And then, when it hasn't finished initializing, the error occurs, because the the application still don't have the reference to the connection to the soap server. And finally notice that it completed initializing the connector after the error. I observed other connectors (mysql, postgresql, rest and they do run well).

<img width="1076" alt="screen shot 2018-07-23 at 22 28 46" src="https://user-images.githubusercontent.com/4633823/43117011-cae45b3c-8ec7-11e8-8835-a3d4675ec5d7.png">

**Comment by shimks** [comment]
I'm working on a fix for this right now. My plan for the fix is to check whether the datasource had finished initializing in `getService` and set up a promisified handler if it hasn't so that the service does not get returned until its datasource has completely been set up.

**Comment by bajtos** [comment]
@shimks isn't this something our soap connector should handle? For example, our data-access methods like `create` or `findById` start by checking whether the datasource has been set up and defer further work otherwise, see e.g. here:

https://github.com/strongloop/loopback-datasource-juggler/blob/4d14c1473fe5a670369553a1969efffe8b05c71a/lib/dao.js#L1336-L1339

Can we reproduce this problem in LoopBack 3.x too? That would be another good reason to fix the problem in the connector, not at LB4 level.

**Comment by shimks** [comment]
@bajtos I agree that this is something the soap connector should handle ideally. I took a look at how we advocate users to get around this issue in our docs [here](https://loopback.io/doc/en/lb3/SOAP-connector.html#creating-a-model-from-a-soap-data-source-1) and thought that this was the way we should handle this issue in LoopBack 4 as well.

Having taken a look at your linked `datasource.ready` function though, it seems like the queue here is not working as intended. I'll try to reproduce this problem in LoopBack 3

**Comment by shimks** [comment]
I should've realized this much earlier before I sank 5 hours into this issue, but I don't think there's a way to verify that the operations exist for the datasource's DAO in between the time the datasource's been initialized and the time the SOAP service proxy has been fully created. The only way to make sure that the service has been hooked up is if we can create a datasource asynchronously, but that seems to require a design overhaul with the juggler.

Considering we required the users to use an event listener for SOAP connector in the first place, I think it's fine for LoopBack 4 as a consumer of the juggler to use a listener as well. What do you think?

**Comment by marioestradarosa** [comment]
Why not put both type of functions to allow developer to choose ?
```javascript
export function getService<T = GenericService>(ds: legacy.DataSource): T {
  // tslint:disable-next-line:no-any
  return ds.DataAccessObject as any;
}

export async function getServiceAsync<T = GenericService>(
  ds: legacy.DataSource,
): Promise<T> {
  await ds.connect();
  // tslint:disable-next-line:no-any
  return ds.DataAccessObject as any;
}
```

**Comment by bajtos** [comment]
> I don't think there's a way to verify that the operations exist for the datasource's DAO in between the time the datasource's been initialized and the time the SOAP service proxy has been fully created. The only way to make sure that the service has been hooked up is if we can create a datasource asynchronously, but that seems to require a design overhaul with the juggler.

Hmm, I think I am starting to understand the problem. So IIUC, this problem affects datasources using dynamic configuration where we asynchronously download WSDL spec from the given URL/file and build DataAccessObject methods from the API described in the downloaded WSDL.

Because the DataAccessObject API depends on the content of WSDL, we cannot create DAO methods in advance as we do for database access in PersistedModel.

@shimks could you please confirm that my understanding is correct?

> Why not put both type of functions to allow developer to choose?

IMO, developers should not have to choose sync/async version depending on the connector and/or connector version(!) they are using. It's adding more work for LB4 app developers to do and more learning to go through. Ideally, there should be a single abstraction (e.g. `async getService(ds)`) that supports all webservice connectors we have.

**Comment by marioestradarosa** [comment]
I confirm I did the following, inspecting @shimks  changes on his PR.

1. Modified the file legacy-juggler-bridge.ts and made it async and force the await of connect
2. Modified calculator.service.ts and add a Promise<CalculatorService> 
3. Modified calculator.service.integration.ts and added await in the new CalculatorServiceProvider() constructor
After modifying these 3 places, it is working as expected.

<img width="832" alt="screen shot 2018-07-25 at 18 15 41" src="https://user-images.githubusercontent.com/4633823/43234094-5724c6de-9037-11e8-9fbd-78a3643e0d47.png">

**Comment by marioestradarosa** [comment]
You are right, the SOAP connector is taking more time to complete the connection, since all the activity that it has to do (locating the module, load it, load and parse wsdl configuration and then create the methods from the datasource.json configuration file). Even on my acceptance test in the calculator I have to override the mocha 2 secs for this reason.

**Comment by shimks** [comment]
@bajtos Yup, I think you've accurately described the problem.

**Comment by marioestradarosa** [comment]
solved with PR #1563
